shared-libs/cht-datasource/src/local/person.ts

```diff
@@ -58,4 +58,22 @@ export namespace v1 {
       return deepCopy(personWithLineage);
     };
   };
+
+  /** @internal */
+  export const getPage = ({ medicDb, settings }: LocalDataContext) => {
+    return async (personType: ContactTypeQualifier, limit: number, skip: number): Promise<Person.v1.Person[]> => {
+      const personTypes = contactTypeUtils.getPersonTypes(settings.getAll());
+      const personTypesIds = personTypes.map(item => item.id);
+
+      if (!personTypesIds.includes(personType.contactType)) {
+        throw new Error(`Invalid person type: ${personType.contactType}`);
+      }
+
+      const getDocsByPage = queryDocsByKey(medicDb, 'medic-client/contacts_by_type');
+
+      const docs = await getDocsByPage([personType.contactType], limit, skip);
+
+      return docs.filter((doc): doc is Person.v1.Person => isPerson(settings, doc, doc?._id));
```

Comment:
@m5r @sugat009 I have realized a complication here while looking at the code over in https://github.com/medic/cht-core/pull/9281.  Basically, we pass the `limit` through to the Couch view query and so we know `docs.length === limit` _or we have reached the last page of results_.  However, before this function returns the persons, we are _filtering_ the array. In theory it is possible that some of the results returned from `queryDocsByKey` could get filtered out (with our current validation logic I cannot come up with a feasible scenario where this could happen, but if/when we add more validations, this will become an issue).  

The reason I think this filtering is a problem is because it could result in us giving back an array for the page that is `< limit` when there are still additional results to retrieve.  Of course, this is bad for our `getAll` logic, but it will also probably be confusing to any other consumers of this code. (How would anyone know when to stop calling `getPage`?).  

The only feasible option I can come up with here is to change the return type to  `Promise<Nullable<Person.v1.Person>[]>` and just pass through `null` when a returned doc fails the `isPerson` check.  If we try to do anything more fancy, like calling the `queryDocsByKey` again until we get the full limit of persons, it is going to mess up the `skip` value for any subsequent calls from the consumer...  The only comfort I have here is that we do not really want folks directly calling this function anyway (and the `getAll` can still abstract away the `Nullable`).  The REST api is going to be a bit janky, but I just don't see any other way around it.

Do either of you have other thoughts here?  Ideas on better options?

Reply:
TBH, I'm surprised that we even need a `isPerson` validator since we explicitly pass a `personType` to the database. Anyways, @jkuester is right, if in some way an entry is filtered by the `isPerson` function then the sanctity of the pagination would be disturbed, so let's have a `null` value in the place of the non-person doc. 

As an alternative approach, instead of setting the document entry to null, we could introduce an invalid property or similar indicator. This would signify that the `person` document contains incorrect or problematic data, without completely removing the entry.

Reply:
>  I'm surprised that we even need a isPerson validator since we explicitly pass a valid personType to the database

This is true. I have gone back and forth on if we should keep the `isPerson` check since all it is doing currently is validating the contact type (which, in theory should always be correct since that is what we are querying one...).  _However_, regardless of if we need to do any checks like this currently, I do think it is very important that we design these API patterns in such a way that we _can_ filter data we get from Couch before returning it to the user.  One of the main points of having cht-datasource is that not everything needs to be modeled directly in the db....  With that in mind, and seeing as the performance cost of `isPerson` is minimal, it makes sense to me to keep it for the sake of code consistency (and better support for future validations).  

>  the sanctity of the pagination would be disturbed

Quote of the day!  :fire: 

> we could introduce an invalid property or similar indicator

This is an interesting idea! :thinking:   Similar to the Pouch API which returns the row objects with `id` value, etc, but may or may not actually have a `doc`.  Ultimately, though, it seems like it might not actually be useful to consumers of this API (and anyone trying to do data validation will probably want to connect directly to the database).

---

Another approach we could take here would be to return some kind of page object like this:

```ts
Promise<{ entries: Person.v1.Person[], cursor: string }>
```

This is similar to the paging in GitHub's GraphQL api, where each request returns a different `cursor` that  you can pass back in on subsequent requests to get the next page.  In our case we could just return the proper `skip` value for the next Couch query, but the meaning/value of the `cursor` should be completely abstracted away from consumers of the cht-datasource api. Consumers can just treat it as a token string for getting the next page.  That would give us flexibility in the future if we needed to change up some underlying details of how things work.  In the context of the above discussion regarding `isPerson`, this approach would let us run followup queries to populate a full list of persons to return so that `entries.length === limit` while returning the proper `skip` value for the next page request as the `cursor`.  

I don't know guys, the more I think about this approach, the more I like it!  It decouples us more from the underlying Couch implementation, which I think is a win, giving us more flexibility for future innovation....

Reply:
One variation on this new page object proposal would be to include a `hasNextPage: boolean` in the returned page object so that consumers know when more data is available.  Then, we could clarify that the provided `limit` would be the _max_ value possible for `entries`, (but some pages might be returned with less).  This would free us from having to implement any kind of followup queries in the `getPage` logic (at least at this time).  We could run the Couch query, do any necessary filtering, and return what is left.

Reply:
@jkuester Cursor-based pagination sounds great and would help if any record is filtered by the `isPerson` or any future filters. As for the implementation, should we do it via:
1.  `overfetching` method where we fetch more docs than the limit specifies such that if records are filtered out then, the over-fetched docs can fill in
1. `re-fetching` method where we re-query the database with the amount of docs that were filtered out.

Usually, it is said to be better to minimize the no. of queries to the database which is even more true in our case. 😃 So, my vote is in the first method. What do you think?

Reply:
Cursor-based pagination always felt a bit technically clunky but they make a lot of sense from a consumer point of view. I found a couple of resources to grok and compare these two types of pagination and these are the most relevant ones for us:
- https://use-the-index-luke.com/no-offset
- https://guide.couchdb.org/draft/recipes.html#pagination

A scary but interesting bit from CouchDB's guide to pagination (related to [this comment](https://github.com/medic/cht-core/blob/master/sentinel/src/lib/purging.js#L404) in our purging code):
> The `skip` argument works like this: in addition to going to the first row and starting to read, `skip` will skip as many rows as specified, but CouchDB will still read from the first row; it just won’t return any values for the skipped rows. If you specify `skip=100,` CouchDB will read 100 rows and not create output for them. This doesn’t sound too bad, but it is *very* bad, when you use `1000` or even `10000` as `skip` values. CouchDB will have to look at a lot of rows unnecessarily. 

That doesn't really answer "what should we do when we have entries that shouldn't be in the response?".
I personally don't support potentially returning `null` entries in the middle of the response when the user requested 100 contacts... I would much rather do some plumbing to fill up the response with the desired amount of contacts but I can only see this approach working with cursor-based pagination

Reply:
Nice! This discussion is great!  (Also, super helpful resources Mokhar, thanks!  :+1: )  I think that definitely seals the deal that we should ditch `skip` and use a `cursor`.  So, `getPage` could return something like:

```ts
export interface Page<T> {
  readonly data: T[];
  readonly cursor: string;
} 
```

With that in mind, I do think it makes sense to put the extra effort in to avoid returning null.  @sugat009 I like your breakdown of `overfetching` vs `re-fetching`. I guess we cannot _only_ rely on overfetching, though, because there might be a case where we end up filtering out a bunch of docs and so we still don't have enough to return (even though we originally fetched extra...).  I am considering a hypothetical future scenario where maybe we start validating that all returned persons have a `name` value.  Docs with `person` type but no name will be returned from the Couch view.  (And maybe you have a CHT instance where someone ran a script to dump a bunch of invalid `person` contacts with no name.)  In that case, we need to be able to deal with disregarding a bunch of those person contacts.

What do you think of an approach where we first just run a view query for the given limit (no overfetch).  Then, we take the returned docs and filter them.  If we ended up filtering any docs (lets say we filtered out `n` docs), then we should re-fetch for the next `n*2` docs.  For our followup query, we are overfetching by a factor of 2, which should significantly reduce the chances of needing to do additional re-fetches.  Here are the benefits I see of this approach:

- Optimizes for the happy-path, good data workflow.  The only thing we should really be filtering for is bad data.  (Any more specific queries for expected good data should be implemented as a new Couch view and not by querying a broader view and filtering the results).  So, hopefully we are discussing edge case behavior with this filtering
- The  "algorithm" for overfetching on the follow-up calls is simple, but still dynamically tailored to what we know about the state of the data being queried.  (If we filtered `5/100`, we only re-query for the next `10`, but if we filtered `75/100`, we know there is a lot of bad data and we instead query for `150` on the re-fetch)

Reply:
I read through the linked sites by @m5r  and tried a bit on our couchdb view(`medic-client/contacts_by_type`). The `startkey` as a cursor approach does not seem to work for the view as all the keys are person types. We probably need to create a new view designed specifically for pagination.

Reply:
I have not tried it yet, but at the end of that Couch guide, it notes that for views where the key is not unique, you will need to use a combination of `startkey`, `startkey_docid`, and `limit`.  The `startkey_docid` is what will actual do the heavy lifting for our contacts_by_type query since we are not actually looking for a _range_ of keys...

Reply:
Good point Josh, the fact that the `key` is the same is what enables couch to do the pagination here. I gave it a try locally and it should do exactly what we need:

Querying the view for persons:
```sh
$ curl "http://admin:password@localhost:5984/medic/_design/medic-client/_view/contacts_by_type?key=\[\"person\"\]"
{"total_rows":6,"offset":3,"rows":[
{"id":"3e93a66c-06c9-4766-b68b-eaec896ee935","key":["person"],"value":"false false 3 grootchw"},
{"id":"6892b0eb-7398-47a9-8e92-d40d3012f9f6","key":["person"],"value":"false false 3 grootvisor"},
{"id":"de324069-da24-4ffa-89f3-f7f67b859d70","key":["person"],"value":"false false 3 grootpatient"}
]}
```

Then using the person with id `6892b0eb-7398-47a9-8e92-d40d3012f9f6` as cursor:
```sh
$ curl "http://admin:password@localhost:5984/medic/_design/medic-client/_view/contacts_by_type?key=\[\"person\"\]&startkey_docid=6892b0eb-7398-47a9-8e92-d40d3012f9f6"
{"total_rows":6,"offset":4,"rows":[
{"id":"6892b0eb-7398-47a9-8e92-d40d3012f9f6","key":["person"],"value":"false false 3 grootvisor"},
{"id":"de324069-da24-4ffa-89f3-f7f67b859d70","key":["person"],"value":"false false 3 grootpatient"}
]}
```

Then limiting the number of rows that I get back:
```sh
$ curl "http://admin:password@localhost:5984/medic/_design/medic-client/_view/contacts_by_type?key=\[\"person\"\]&startkey_docid=6892b0eb-7398-47a9-8e92-d40d3012f9f6&limit=1"
{"total_rows":6,"offset":4,"rows":[
{"id":"6892b0eb-7398-47a9-8e92-d40d3012f9f6","key":["person"],"value":"false false 3 grootvisor"}
]}
```

Reply:
Ahhh! Gotcha. I was under the impression that the `key` also needed to be a uuid. Thanks!

Reply:
Okay, so just recapping further discussion from the [Slack Thread](https://medic.slack.com/archives/C024KTGRW/p1722374179017329), it turns out that Pouch does not have any support for the `startkey_docid` parameter.  :sweat:   Also from reading the [Couch docs](https://docs.couchdb.org/en/stable/ddocs/views/pagination.html#paging-alternate-method) it seems like the performance for `skip` significantly improved after Couch `1.2` (I guess the guide linked above was old?).  The [Pouch Docs](https://pouchdb.com/api.html#query_database) still note that `skip` has "poor performance on IndexedDB/LevelDB", but there does not seem to be any other way to actually page queries with Pouch, so... :shrug:

So the decision is to stick with just using `skip` to page the queries for now.  We should still do the refactor to return the skip value as a generic `cursor` to consumers instead of actually exposing `skip`.  That will let us passively refactor our implementation once https://github.com/pouchdb/pouchdb/issues/7326 is completed.